src/ArgTools.jl

```diff
@@ -62,14 +62,32 @@ The `arg_read` function accepts an argument `arg` that can be any of these:
 Whether the body returns normally or throws an error, a path which is opened
 will be closed before returning from `arg_read` and an `IO` handle will be
 flushed but not closed before returning from `arg_read`.
+
+If `lock=false` is passed, then any `IOStream` will be used without locking,
+which can be a significant performance improvement. You should, however, only
+turn locking off if `f` uses its I/O argument in a threadsafe manner.
 """
-arg_read(f::Function, arg::AbstractString) = open_nolock(f, arg)
-arg_read(f::Function, arg::ArgRead) = open(f, arg)
-arg_read(f::Function, arg::IO) = f(arg)
+arg_read(f::Function, arg::AbstractString; lock::Bool=true) =
+    open_lock(f, arg, lock=lock)
+arg_read(f::Function, arg::ArgRead; lock::Bool=true) = open(f, arg)
+arg_read(f::Function, arg::IO; lock::Bool=true) = f(arg)
+
+VERSION ≥ v"1.5" &&
+function arg_read(f::Function, arg::IOStream; lock::Bool=true)
+    arg_dolock = arg._dolock
+    arg._dolock = lock
+    try # take outer lock if !lock
+        lock || Base.lock(arg)
+        f(arg)
+    finally
+        lock || unlock(arg)
+        arg._dolock = arg_dolock
+    end
```

Comment:
This implementation seems to be nearly identical to the default one in Base, but appears to introduce several unnecessary data races on the `arg._dolock`  variable. But, falling back on the default behaviors should be possible to optimize to give the same performance.
```suggestion
    # confusing, lock=true means `f` is potentially multi-threaded (therefore locking here is bad, and will also disable finalizers)
    # while lock=false means it is single-threaded (so locking here is good hygiene, though it may disable finalizers)
    lock || Base.lock(arg)
    try
        f(arg)
    finally
        lock || unlock(arg)
    end
```

Reply:
I don't really understand the comments about finalizers. How would this affect them?

Reply:
finalizers cannot run while any code holds a lock (on that Task)